app/src/main/java/com/skyous/trakkit/ui/navigation/core/MainNavigator.kt

```diff
@@ -0,0 +1,227 @@
+     * is the same as the currently displayed, false otherwise
+     */
+    fun navigateTo(tab: MainNavigator.Tab): Boolean {
+        return navigateToInternal(tab, null, false)
+    }
+
+    /**
+     * Navigates to the {@see MainNavigator.Tab} with the given resource ID
+     *
+     * @return true if the given tab already exists in the nav history or if the given tab
+     * is the same as the currently displayed, false otherwise
+     */
+    fun navigateTo(tabResourceId: Int): Boolean {
+        return navigateToInternal(MainNavigator.Tab.fromItemResourceId(tabResourceId), null, false)
+    }
+
+    /**
+     * Navigates to the given {@see MainNavigator.Tab}.
+     *
+     * @param extras extras to be set to the destination tab's fragment
+     *
+     * @return true if the given tab already exists in the nav history or if the given tab
+     * is the same as the currently displayed, false otherwise
+     */
+    fun navigateTo(tab: MainNavigator.Tab, extras: Bundle?): Boolean {
+        return navigateToInternal(tab, extras, false)
+    }
+
+    /**
+     * Navigates to the MainNavigator.Tab with the given resource ID
+     *
+     * @param extras extras to be set to the destination tab's fragment
+     *
+     * @return true if the given tab already exists in the nav history or if the given tab
+     * is the same as the currently displayed, false otherwise
+     */
+    fun navigateTo(tabResourceId: Int, extras: Bundle?): Boolean {
+        return navigateToInternal(MainNavigator.Tab.fromItemResourceId(tabResourceId), extras, false)
+    }
+
+    /**
+     * Function to be called from host hostActivity's onBackPressed method.
+     *
+     * @return true if the back navigation was handled, false otherwise
+     * False value means that host should handle back navigation
+     */
+    fun onBackPressed(): Boolean {
+        return when {
+            navigationHistory.size > 1 -> {
+                val tabToNavigate = navigationHistory[navigationHistory.lastIndex - 1]
+                Log.d(TAG, "Navigating back...")
+                navigateToInternal(tabToNavigate, null, true)
+                backNavigationListener?.onNavigatedBack(tabToNavigate)
+                true
+            }
+            else -> false
+        }
+    }
+
+    /**
+     * If the navigator should save it's state before process death, this
+     * hostActivity must be called, passing the outState bundle.
+     */
+    fun onSaveInstanceState(outSate: Bundle?) {
+        Log.d(TAG, "Saving navigation history:" + navigationHistory.toString())
+        outSate?.putSerializable("nav_history", navigationHistory)
+    }
+
+    fun onDestroy() {
+        navigationHistory.clear()
+        fragmentContainer.clear()
+        backNavigationListener = null
+        hostActivity.clear()
+    }
+
+    private fun navigateToInternal(tab: MainNavigator.Tab, extras: Bundle?, isBackNavigation: Boolean): Boolean {
```

Comment:
would be nice to have a navigateToInternal method without a bundle param, since you are calling this and passing null values in multiple places

Reply:
or it would be nicer to have named params, then it one signature is enough I guess